zds/tutorialv2/api/serializers.py

```diff
@@ -0,0 +1,163 @@
+
+    def __init__(self, *args, **kwargs):
+        self.db_object = kwargs.pop('db_object', None)
+        super(ChildrenListSerializer, self).__init__(*args, **kwargs)
+        self._validated_data = {}
+        self._errors = {}
+
+    def create(self, validated_data):
+        return ChildrenListViewModel(**validated_data)
+
+    def save(self, **kwargs):
+        self.is_valid(True)
+        return self.create(self.validated_data)
+
+    @transform(TypeError, ValidationError, 'incorrect json')
+    def is_valid(self, raise_exception=False):
+        """
+        This method overrides ``ModelSerializer`` method. If someone knows a cleaner way, dig it.
+
+        :param raise_exception:
+        :return:
+        """
+
+        error = not super(ChildrenListSerializer, self).is_valid(raise_exception)
+        messages = {}
+
+        for field_name, value in self.initial_data.items():
+            if field_name in self.Meta.fields:
+                self._validated_data[field_name] = value
+        if self._validated_data.get('extracts', None):
+            self._validated_data['extracts'] = [ChildrenViewModel(**v) for v in self._validated_data['extracts']]
+        if self.initial_data.get('containers', None):
+            self._validated_data['containers'] = [ChildrenViewModel(**v) for v in self._validated_data['containers']]
+        if not all(c.child_type.lower() == 'extract' for c in self._validated_data.get('extracts', [])):
+            error = True
+            messages['extracts'] = _('un extrait est mal configuré')
+        if len(self._validated_data['extracts']) != len(set(e.title for e in self._validated_data['extracts'])):
+            error = True
+            titles = Counter(list(e.title for e in self._validated_data['extracts']))
+            doubly = [key for key, v in titles.items() if v > 1]
+            messages['extracts'] = _('Certains titres sont en double : {}').format(','.join(doubly))
+        if len(self._validated_data['containers']) != len(set(e.title for e in self._validated_data['containers'])):
+            error = True
+            titles = Counter(list(e.title for e in self._validated_data['containers']))
+            doubly = [key for key, v in titles.items() if v > 1]
+            messages['containers'] = _('Certaines parties ou chapitres sont en double : {}').format(','.join(doubly))
+        if not all(c.child_type.lower() == 'container' for c in self._validated_data.get('containers', [])):
+            error = True
+            messages['containers'] = _('Un conteneur est mal configuré')
+        self._validated_data['introduction'] = self.initial_data.get('introduction', '')
+        self._validated_data['conclusion'] = self.initial_data.get('conclusion', '')
+        if not self._validated_data['extracts'] and not self._validated_data['containers']:
+            error = True
+            messages['extracts'] = _('Le contenu semble vide.')
+        if raise_exception and error:
```

Comment:
J’ai un peu du mal à voir l’utilité de la variable `error`. On ne peut pas faire `if raise_exception and messages != {}:` ?

Reply:
on peut, j'essayais de me dépatouiller avec le fait que je n'utilise pas un sérializer standard car ce n'est pas un modèle en bdd.